### PR TITLE
Scalaz 8 typeclass encoding. WIP.

### PR DESCRIPTION
**Pull request by kamilkloch**
First draft towards https://github.com/typelevel/cats-mtl/issues/26.
New encodings for:
- [x] ApplicativeAsk
- [x] ApplicativeLocal
- [x] MonadChronicle

### CODE REVIEW

File: core/src/main/scala/cats/mtl/ApplicativeAsk.scala [hunk 1]
```diff
@@ -1,5 +1,8 @@
-package cats
-package mtl
+package cats.mtl
+
+import cats.Applicative
+
+import acyclic.skipped
```
Comment (LukaJCB) [inline core/src/main/scala/cats/mtl/ApplicativeAsk.scala]: What does this do? Do we need it? :)

Comment (kamilkloch) [reply core/src/main/scala/cats/mtl/ApplicativeAsk.scala]: Result of `scalacOptions += "-P:acyclic:force"`

**Comment by LukaJCB** [comment]
Looks good so far, thank you Kamil! 
I wonder if we could get some good benchmarks to really motivate this :)

**Comment by kamilkloch** [comment]
@LukaJCB please take a look into https://gitter.im/scalaz/scalaz, @edmundnoble shed some light.

**Comment by kamilkloch** [comment]
@LukaJCB You are right. This would simplify the code at the cost of adding a highly unlikely potential implicit ambiguity. I was thinking of yet another thing: is there any way to actually inherit (not aggregate) from cats' typeclasses? I mean `ApplicativeAsk[F[_], E] extends Applicative[F]` ;)

**Comment by SystemFw** [comment]
>  is there any way to actually inherit (not aggregate) from cats' typeclasses? I mean ApplicativeAsk[F[_], E] extends Applicative[F]

No, that would bring you the ambiguous implicit problem `scato` solves, since e.g. `F[_]: ApplicativeAsk : ApplicativeLocal` would have two `Applicative`s

**Comment by LukaJCB** [comment]
We could use the `instanceOf` pattern for `ApplicativeAsk` and `Applicative`, though I'm sceptical that would buy us all too much.

**Comment by LukaJCB** [comment]
> We can do that, but it kinda constrains to not add any more leaves right? Unsure if it's a problem or not, practically

@SystemFw if we do add more leaves, we can just `scato` for that leaf again.

**Comment by LukaJCB** [comment]
> F[_]: ApplicativeAsk : ApplicativeLocal would also fail if ApplicativeLocal would plainly extend ApplicativeAsk, as suggested by @LukaJCB . I was more thinking of using the hidden inheritance pattern from scalaz also for Applicative, for example
ApplicativeAskClass[F[_], E] extends ApplicativeClass[F]
(i.e., switch fully to the ineritance mode by adding some instanceOf plumbing, just an idea)

I'm not sure if this buys us a lot, unless we provide an implicit conversion from e.g. `ApplicativeAsk` to `Applicative`. Because right now, `cats-mtl` has you write code like this:

```scala
def foo[F[_]: Monad: ApplicativeAsk[?[_], Config]: F[A] = ???
```

**Comment by LukaJCB** [comment]
Yes, I think it looks good this way :+1:

**Comment by kamilkloch** [comment]
If we are to proceed, shouldn't we do it in a new PR?

**Comment by LukaJCB** [comment]
Sure Sounds good to me :)

**Comment by LukaJCB** [comment]
I'll close this for now, we can always reopen if we reconsider :)

**Pull request closed without merging**